hyperspy/signal.py

```diff
@@ -3288,99 +3288,227 @@ def save(
+                    If set to True the signal or first signal in the list of signals
+                    will be defined as the default (following Nexus v3 data rules).
+                write_dataset : bool, optional
+                    Only for hspy files. If True, write the dataset, otherwise, don't
+                    write it. Useful to save attributes without having to write the
+                    whole dataset. Default is True.
+                close_file : bool, optional
+                    Only for hdf5-based files and some zarr store. Close the file after
+                    writing. Default is True.
+                file_format: string
+                    The file format of choice to save the file. If not given, it is inferred
+                    from the file extension.
+
+                Examples
+                --------
+                Save with a complete filename:
+
+                >>> s = hs.signals.Signal1D(np.arange(10))
+                >>> s.save("my_data.hspy")
+
+                Re-save a loaded signal in a different location (uses tmp_parameters):
+
+                >>> s = hs.load("original_data.hspy")  # tmp_parameters are auto-populated
+                >>> s.save("/new/output/folder/")      # Saves to /new/output/folder/original_data.hspy
+
+                Re-save a loaded signal in a different format:
+
+                >>> s = hs.load("data.hspy")           # Original format
+                >>> s.save("/output/", file_format="msa")  # Convert to MSA format
+                >>> s.save("/output/", file_format="rpl")  # Convert to Ripple format
+
+                Process multiple files and save in different format:
+
+                >>> import hyperspy.api as hs
+                >>> from pathlib import Path
+                >>>
+                >>> input_folder = Path("input_data/")
+                >>> output_folder = Path("processed_data/")
+                >>>
+                >>> for file_path in input_folder.glob("*.hspy"):
+                ...     s = hs.load(file_path)
+                ...     # Process the signal...
+                ...     s = s.remove_background()
+                ...     # Save in new location with different format
+                ...     s.save(output_folder, file_format="msa")
 
         """
+        # Check for conflicting parameters
+        if extension is not None and file_format is not None:
+            raise ValueError(
+                "Cannot specify both 'extension' and 'file_format' parameters. "
+                "Please use only 'file_format' as 'extension' is deprecated."
+            )
+
+        # Deprecation warning for extension parameter
+        if extension is not None:
+            warnings.warn(
+                "The 'extension' parameter is deprecated in HyperSpy 2.4 and will be removed in HyperSpy 3.0. "
+                "Please use 'file_format' instead.",
+                FutureWarning,
+                stacklevel=2,
+            )
```

Comment:
In the interest of simplify the logic in the rest of this function (and also when it comes to remove the deprecation), would it be better to do the parsing of the `extension` to the `file_format` argument - assuming this is possible?

Reply:
I didn't take that option because in corner cases, it is not exactly equivalent. For example, if you want to use an extension that is not the default extension for the file format.

Reply:
It means that the support for these corner cases will be dropped in HyperSpy 3.0? Is it something that we want, should be documented (comment in the code to be used when removing the functionality so that this API break can be documented in the 3.0 release)?